### PR TITLE
Basic DocumentHighlightKind support for assignments

### PR DESCRIPTION
**Pull request by kjeremy**
Wraps references per #2738 and adds limited support for DocumentHighlightKind Read/Write for simple binops assignments.

I think I need some help with determining reads/writes.

Towards #2560

### CODE REVIEW

**Comment by kjeremy** [comment]
This change enables vscode to highlight variable assignment differently from reads.

**Comment by kjeremy** [comment]
Handles basic writes for `=`

**Comment by matklad** [comment]
I've merged the `StructLiteral` PR first, so this needs to be rebased.

**Comment by kjeremy** [comment]
@matklad done

File: crates/ra_ide/src/references.rs [hunk 1]
```diff
@@ -210,11 +224,50 @@ fn process_definition(
...
 }
 
+fn access_mode(kind: NameKind, name_ref: &ast::NameRef) -> Option<ReferenceAccess> {
+    match kind {
+        NameKind::Local(_) | NameKind::Field(_) => {
+            //LetExpr or BinExpr
+            name_ref.syntax().ancestors().find_map(|node| {
+                match_ast! {
+                    match (node) {
+                        ast::BinExpr(expr) => {
```
Comment (matklad) [inline crates/ra_ide/src/references.rs]: If we only look at `BinExpr`, it's simpler to `.filter_map(ast::BinExpr::cast)`

Comment (kjeremy) [reply crates/ra_ide/src/references.rs]: `LetExpr` is on the TODO list and I should update the comment above. Does `filter_map` still make  sense? I think I'd just have to change it later.

Comment (matklad) [reply crates/ra_ide/src/references.rs]: Hm, yeah, `match_ast` makes sense here. That said, the rightward drift seems non-trivial here. Perhaps  something can be factored into a function? 

Or maybe we can just use an early return for the outer match?

Comment (kjeremy) [reply crates/ra_ide/src/references.rs]: In thinking about it... if we're in a `ast::BinExpr` and not in the lhs of an assignment we can guarantee we're in a read so I think I can change that inner bit to:

```rust
 ast::BinExpr(expr) => {
	if expr.op_kind()?.is_assignment() {
		// If the variable or field ends on the LHS's end then it's a Write (covers fields and locals).
		// FIXME: This is not terribly accurate.
		if let Some(lhs) = expr.lhs() {
			if lhs.syntax().text_range().end() == name_ref.syntax().text_range().end() {
				return Some(ReferenceAccess::Write);
			}
		}
	}
	
	return Some(ReferenceAccess::Read);
},
```

**Comment by kjeremy** [comment]
I can't figure out how to move this back to draft status but I'm going through the tests and adding new cases.

**Comment by kiljacken** [comment]
That's not supported by GitHub (hopefully "yet"), so that would explain why you can't find it. Ran into the same issue at some point..

**Comment by kjeremy** [comment]
@matklad I think this is ready for another pass.

**Comment by kjeremy** [comment]
Notice how the `let text` is highlighted in a way that shows it is a write to that symbol (cursor is on the last usage).

image

**Comment by kiljacken** [comment]
That looks pretty useful! Good job

File: crates/ra_syntax/src/ast/expr_extensions.rs [hunk 2]
```diff
@@ -144,6 +144,7 @@ impl BinOp {
         }
     }
 }
+
```
Comment (matklad) [inline crates/ra_syntax/src/ast/expr_extensions.rs]: :+1:

**Pull request merged**